### PR TITLE
Add a test for gl.uniform*() with optional args: srcOffset / srcLength.

### PR DESCRIPTION
**Pull request by zhenyao**
Tested with Firefox and all test cases pass.

### CODE REVIEW

**Comment by zhenyao** [comment]
@kainino0x @kenrussell @jdashg PTAL

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+  var typeInfo = typeInfos[tt];
+  debug("");
+  debug("check " + typeInfo.type);
+  var srcLength = typeInfo.srcValues.length;
+  var fSrc = wtu.replaceParams(fTemplate, typeInfo);
+  //debug("fSrc: " + fSrc);
+  var program = wtu.loadProgram(gl, vSrc, fSrc);
+  if (!program) {
+    testFailed("Failed to create the program");
+    alert(fSrc);
```
Comment (qkmiao) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: The debug code can be removed.

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+  }
+}
+
+for (var tt = 0; tt < typeInfos.length; ++tt) {
+  var typeInfo = typeInfos[tt];
+  debug("");
+  debug("check " + typeInfo.type);
+  var srcLength = typeInfo.srcValues.length;
+  var fSrc = wtu.replaceParams(fTemplate, typeInfo);
+  //debug("fSrc: " + fSrc);
```
Comment (qkmiao) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: The debug code can be removed.

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+  },
+];
+
+for (var tt = 0; tt < typeInfos.length; ++tt) {
+  var typeInfo = typeInfos[tt];
+  debug("");
+  debug("check " + typeInfo.type);
+  var srcLength = typeInfo.srcValues.length;
+  var fSrc = wtu.replaceParams(fTemplate, typeInfo);
+  //debug("fSrc: " + fSrc);
```
Comment (qkmiao) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: same here

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+  var program = wtu.loadProgram(gl, vSrc, fSrc);
+  if (!program) {
+    testFailed("Failed to create the program");
+    alert(fSrc);
```
Comment (qkmiao) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: same here.

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html (hunk 2 above)
Comment (zhenyao) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: Done.  Here and below.

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+      " with srcOffset = 0 / srcLength = 0");
+  checkUniformValues(program, typeInfo);
+
+  gl[typeInfo.setter](loc, typeInfo.srcValuesWithOffset, typeInfo.srcOffset);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+      "can set an array of uniforms with gl." + typeInfo.setter +
+      " with srcOffset = " + typeInfo.srcOffset);
+  checkUniformValues(program, typeInfo);
+
+  gl[typeInfo.setter](loc, typeInfo.srcValuesWithOffset, typeInfo.srcOffset);
```
Comment (kenrussell) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: This test looks redundant with respect to the one just above it?

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+      " with srcOffset = 0 / srcLength = 0");
+  checkUniformValues(program, typeInfo);
+
+  gl[typeInfo.setter](loc, gl.FALSE, typeInfo.srcValuesWithOffset, typeInfo.srcOffset);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+      "can set an array of uniforms with gl." + typeInfo.setter +
+      " with srcOffset = " + typeInfo.srcOffset);
+  checkUniformValues(program, typeInfo);
+
+  gl[typeInfo.setter](loc, gl.FALSE, typeInfo.srcValuesWithOffset, typeInfo.srcOffset);
```
Comment (kenrussell) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: This one looks redundant too. A mistake?

File: sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html (hunk 5 above)
Comment (zhenyao) [inline sdk/tests/conformance2/uniforms/gl-uniform-arrays-sub-source.html]: Thanks for catching it.  Removed (here and below)

**Comment by zhenyao** [comment]
Since two people already reviewed it, I am going to merge myself.

@kainino0x @jdashg If you have further comments, I'll address them in a follow up.

**Comment by kenrussell** [comment]
Great work Mo. Thanks again for adding these important missing tests.

**Pull request merged**